download/download.go

```diff
@@ -97,30 +101,62 @@ func (d *Downloader) ClearCache() {
 
 // Start tells the Downloader to begin downloading bundles.
 func (d *Downloader) Start(ctx context.Context) {
-	go d.loop()
+	go d.doStart(ctx)
 }
 
-// Stop tells the Downloader to stop begin downloading bundles.
+func (d *Downloader) doStart(ctx context.Context) {
+	ctx, cancel := context.WithCancel(context.Background())
+
+	d.wg.Add(1)
```

Comment:
💭 Is there ever more than one entry in this waitgroup? That is, are there multiple bundles downloaded (or long-polled) in parallel? If so, I guess I don't it...

If we don't need it for multiple entries, I suspect there might be some room for simplification here... we've got so many synchronisation stuff here...

Reply:
We wait for the goroutine that performs the actual download so that it can finish calling the callback function provided by the plugins. For example, in the bundle plugin, after we stop a bundle's downloader, we delete its status object. So if we don't wait for the goroutine to finish, it's possible that plugin may have already deleted the bundle's status object which would result in a panic in the callback function.

Reply:
Thanks, I see. What I still don't understand is if that waitgroup ever has more than 1 group member...? But I'll have another closer look tomorrow.